src/utils.rs

```diff
@@ -6,6 +6,60 @@ use seahash::hash;
 #[cfg(target_pointer_width = "32")]
 use seahash::reference::hash;
 
+/// A stack-allocated vector that uses [T; MAX_SIZE] with Default initialization.
+/// All elements are initialized to T::default(), and we track the logical size separately.
+/// Note: a future impl can switch to using MaybeUninit with unsafe code for better efficiency.
+pub struct ArrayVec<T, const MAX_SIZE: usize> {
+    data: [T; MAX_SIZE],
+    size: usize,
+}
```

Comment:
Let's use one of the existing `tinyvec`, `smallvec`, or `arrayvec` crates for this purpose - they have already implemented, optimized, and tested this kind of struct in production for a long time now.

`tinyvec` and `smallvec` are already in the dep tree for `adblock-rust`; `arrayvec` is also in the dep tree for `brave-core`.

Reply:
Nice catch!
I though we don't use `arrayvec` in b-c, so adding extra deps is undesirable.
Done.